### ISSUE TITLE
HMSL for Amiga is unavailable

### ISSUE THREAD

**Issue by philburk** [philburk/hmsl#166]
Some users are trying to run HMSL on the new Amigas and Amiga emulators.
Files are scattered and some are missing.

JForth for the Amiga can be downloaded from: https://www.softsynth.com/jforth/download.html

Amiga source files are in https://github.com/philburk/hmsl/tree/master/hmsl/amiga

The "Amiga Supplement" mentioned in the main HMSL manual seems to be missing!
The supplement was probably written on the Mac because we used Word and the Apple laser printer.

The file for loading HMSL on top of JForth is at
https://github.com/philburk/hmsl/blob/master/hmsl/amiga/hostdep/load_hmsl.fth

Note that it does not have the ".fth" suffix on the file names.

It may be possible to recover HMSL on Amiga by:
1. Run JForth on Amiga
2. Expand the dictionary
3. Update the load_hmsl.fth to use the ".fth" suffix. Maybe extend this [load file](https://github.com/philburk/hmsl/blob/master/hmsl/fth/load_hmsl.fth).
4. Compile HMSL
5. SAVE-FORTH

**Comment by znmeb** [comment]
There are new Amigas? I knew about the emulators but I didn't know you could get newly-built hardware.

**Comment by philburk** [comment]
I think the new hardware is basically a turnkey dedicated emulator:
http://www.a-eon.com/?page=x5000

**Comment by 320x200** [comment]
(follow-up to my email)

Hello!

So for now I'm just trying a couple of things just to see how far I can go, and I have two questions:
- is it a correct assumption that HMSL for Amiga, should be in fact a combination of platform specific files (`hmsl/amiga/hostdep`) with platform agnostic files (`hmsl/fth`)?
- I have assigned the former to `hh:` and the later to `h:` as well as edited `hh:load_hmsl.fth` so that it points to the needed files with right extension. These assignments and edits seem to help progress in the parsing of the file. I am making changes every time I block somewhere new.
- I had to force `.need HOST=AMIGA` because the test for the `DOS_LIB` constant is failing, which makes me think that a couple of preparatory steps are missing.
- I noticed that `ho:utils` does not seem to exist anywhere, so I replaced it with `h:utils.fth` but then I am stuck because JForth does not know `?DO` so `include load_hmsl.fth` breaks at this point. Could it be that `h:utils.fth` was written for PForth and not JForth? And what is/was `ho:`?

**Comment by philburk** [comment]
The files in "hmsl/amiga/hostdep" are designed to load on top of JForth.
The files in "hmsl//fth" are designed to load on top of pForth.
Those files need to be reconciled.

One problem is that the  "hmsl/amiga/hostdep" is trying to load "hmsl//fth" files on top of JForth.
JForth and pForth are very similar at a basic level. But there are a few differences, eg. DOS_LIB.
You seem to be hitting those.

I have asked the HMSL list if anyone has old Amiga HMSL disks. Fingers crossed.

> JForth does not know ?DO 

I think the JForth was the same as ?DO. Try this in JForth:

0 0 DO i . cr LOOP ." yay"

If it hangs then I am wrong. If it prints "yay" then I am right.
If so you can define an alias. Try:

    : ?DO postpone do ; immediate

> And what is/was ho:?

I think that was the ODE directory.

**Comment by 320x200** [comment]
> > And what is/was ho:?
> 
> I think that was the ODE directory.

Yes! It was the utils from ODE that was missing, not the one from HMSL. However further down the compilation, the `?DO` issue appeared again via `stacks.fth`

> 0 0 DO i . cr LOOP ." yay"

This one `: YAY 0 0 DO i . cr LOOP ." yay" ;` worked. I do get a `"yay"` back. However:

> ```
> : ?DO postpone do ; immediate
> ``

Does not work.
`POSTPONE` does not seem to exist or be defined in JForth?

> I have asked the HMSL list if anyone has old Amiga HMSL disks. Fingers crossed.

Thanks, fingers crossed indeed :)

**Comment by philburk** [comment]
> POSTPONE does not seem to exist or be defined in JForth?

Right. I spent some time today installing an Amiga emulator.
I managed to get JForth running. 
Try this:

    ALIAS DO ?DO IMMEDIATE
    : YAY 0 0 ?DO i . cr LOOP ." yay" ;
    YAY

BTW, The word DO does not work in JForth outside of a colon definition. It works in pForth.

I will help to get HMSL running on the Amiga again.

**Comment by 320x200** [comment]
More progress:
- the `?DO` alias works
- `stacks.fth` failed later because there is a conditional compilation at the end with `[IF]...[THEN]` and it looks like JForth does not support this. It seems it was just some test words, so I commented out the whole block.
- now stuck at the include of `h:midi_globals.fth` that fails right early on with an undefined `hostmidi_port()`. Could it be that `h:midi_globals.fth` is now only meant to work on Mac?

**Comment by philburk** [comment]
JForth has .IF and .THEN
pForth uses [IF] and [THEN], which are more standard.

In h:midi_globals.fth you can see that I had to move the variable into the host code.

```
\ As of 2021, the variable now resides in the Host code.
\ VARIABLE MIDI-PORT      \  Current port being used; 0=modem, 1=printer
: MIDI-PORT hostmidi_port() ;
```

You can reenable the old variable for now.

**Comment by philburk** [comment]
> Could it be that h:midi_globals.fth is now only meant to work on Mac?

The new code does only work on the new Mac. But now I think we should preserve compatibility with JForth. I did not anticipate building HMSL on Amiga from this code base.

**Comment by 320x200** [comment]
> The new code does only work on the new Mac. But now I think we should preserve compatibility with JForth. I did not anticipate building HMSL on Amiga from this code base.

I can imagine... and really appreciate your help to try bringing Amiga support back!

With the `MIDI-PORT` variable back, things moved further. However there were a lot of words used to initialize MIDI for JUCE... I think? I commented out these words in the hope they were not used at all on Amiga.

Now I am stuck here https://github.com/philburk/hmsl/blob/7a2925e1f052dcf8d81283fd0fe86d116c9a08c0/hmsl/amiga/hostdep/load_hmsl.fth#L121

and not quite sure how to solve this because this file does not seem to exist anymore.

**Comment by philburk** [comment]
Oh! That's not good. The event_buffer is code that would store CFAs and data in a time ordered buffer.
This was used for scheduling MIDI and Amiga Local sound events in the near future to reduce jitter.
Scheduling jitter was a big issue on 8 MHz CPUs.

The event buffer functionality in the current HMSL/JUCE is built in the native C code. See this host API:

https://github.com/philburk/hmsl/blob/7a2925e1f052dcf8d81283fd0fe86d116c9a08c0/native/juce/Source/hmsl_host.h#L98

So yeah, those mising files re not needed on the JUCE version but are needed on the Amiga.
I will look again on my old drives.  And I just thought of someone else who may have the HMSL Amiga disks!

**Comment by philburk** [comment]
Good news! I found some files that I got from Henry Lowengard.
It seems to have the original HMSL source including "event_buffer" and others mixed in with some of Henry's stuff.
It might even have an executable hmsl4th!
I will figure out what I have and talk to Henry. I don't want to post his personal files without permission.

**Comment by mdbergmann** [comment]
> I think the new hardware is basically a turnkey dedicated emulator: http://www.a-eon.com/?page=x5000

This is a PowerPC based Amiga. Amiga NG.

But there are also new "classic" FPGA based Amigas like the Apollo V4 (http://www.apollo-computer.com/v4standalone.php) and a vibrating set of new accelerators for old hardware, some even based on Raspberry Pis (PiStorm) which basically emulate a Motorola 68k CPU directly on the Raspi (without host OS) by also passing through Hdmi,Wifi, RAM and SD card, etc.
https://pistorm.github.io

Actually I was on the search for source code of JForth. But couldn't find it. Was it not released?

**Comment by philburk** [comment]
@mdbergmann - The V4 looks nice.

PiStorm sound good but the docs are a bit lacking. https://pistorm.github.io/docs/

There is some JForth source code on the JForth disk images here:
https://www.softsynth.com/jforth/download.html

But I don't think that includes the kernel source, which is 68000 assembly.

**Comment by mdbergmann** [comment]
Hmm, yeah. The PiStorm docs are very rudimentary. Though hardware is already available and working since many months. Maybe those pages contain more info:

https://github.com/michalsc/Emu68 (the 68k jit running directly on ARM)

https://github.com/captain-amygdala/pistorm ( the hardware )

Anyway, sorry for intercepting this ticket. Yeah, Martin Randall spent a lot of efforts getting JForth out as freeware (jforth.org). But shouldn't the kernel sources also be captured?

**Comment by philburk** [comment]
> shouldn't the kernel sources also be captured?

Ideally yes. I am trying to archive as much as I can. But some code was lost along the way. Mike is currently looking for the kernel source.

**Comment by just-me-mail-me** [comment]
Martin Randall here.

Just to clarify things...

I didn't have anything to do with the decision to " getting JForth out as freeware".  That was all Mike Haas and Phil Burk and they had already decided to release it.

I just volunteered to do some legwork in advertizing the release, converting the docs from RTF to PDF, HTML &  WordWorth formats plus create the JForth.org website for posterity, as I think all Amiga stuff should be saved.

Every now and again, I search the web for more JForth information to add to the website, which is how I came across these posts on Github.

Yes, the Apollo Vampire cards and Standalone units like the V4SA and A6000, are great, true Amiga advances both in CPU, DMA, Graphics etc.

https://www.apollo-computer.com/index.php

Worth a look if you have fond memories of the Amiga.
